### PR TITLE
[v1][WIP] Metrics & Stats prototype

### PR DESCRIPTION
**Pull request by rickyyx**
## Requirements and Goals
1. Stats collection should work with both the legacy and new multiprocess architecture.
2. Stats collection should be possible when engine core is busy in doing GPU work (or stuck at something) so we get observability metrics even when engine core is stuck.
3. Stats collection should take into account the entire lifecycle of a request, i.e. the input processing time.
5. Stats collection should incur minimal overheads on runtime/memory --> should likely be decoupled from the engine core loop. 

## Proposal and Design
1. Introducing `RequestStats` and `RequestStatsUpdate` to capture requests updates for metrics. It captures the entire request lifecycle (arrival -> input process -> engine core prefill/decode -> [preempt] -> detokenized). 
2. Introducing a `EngineStatsAgent` that lives with the `EngineCore` to collect stats updates from various core components and scheduler/model outputs
6. Introducing a `EngineStatsManager` class that lives on the `AsyncLLM` engine that aggregates stats from the agent.
7. The manager and agent communicates through the ZMQ sockets for async stats collection. 

```
Main Process                                  Engine Core Process
+----------------------+                      +----------------------+
|      AsyncLLM        |                     |    EngineCoreProc    |
|  +-----------------+ |                     |  +-----------------+  |
|  |EngineStatsManager| |                    |  |EngineStatsAgent |  |
|  +-----------------+ |                     |  +-----------------+  |
|                      |                     |                      |
|  [AsyncIO Event Loop]|     ZMQ Sockets     |  [Background Threads]|
|  +-----------------+ |                     |  +-----------------+  |
|  | Output Handler  |<|---------------------|->| Output Socket   |  |
|  +-----------------+ |                     |  +-----------------+  |
|                      |                     |                      |
|  +-----------------+ |                     |  +-----------------+  |
|  | Stats Handler   |<|---------------------|->| Stats Socket    |  |
|  +-----------------+ |                     |  +-----------------+  |
|                      |                     |                      |
|                      |                     |  +-----------------+  |
|                      |---------------------|->| Input Socket    |  |
|                      |                     |  +-----------------+  |
|                      |                     |                      |
|                      |                     |  [Main Thread]       |
|                      |                     |  +-----------------+  |
|                      |                     |  |   Busy Loop    |  |
|                      |                     |  +-----------------+  |
+----------------------+                      +----------------------+
```

- **On Backend (EngineCore Process)**:
  - New `EngineStatsAgent` collects metrics in a thread-safe manner
  - Stats are sent to frontend via ZMQ sockets
  - Runs in separate background threads to minimize impact on model execution

- **On Frontend (Main Process)**:
  - `EngineStatsManager` aggregates stats from backend as well as collect stats on the engine frontend.
  - AsyncIO-based stats polling with configurable interval
  - Compatible with existing V0 StatsLogger interface 

## Benchmark Results
TLDR: There's no observable perf regression when stats collection happens every 1 sec for the below workloads comparing to stats collection turned off. 

### Throughput 
- PR with stats logging every 1 sec: 13.7 requests/s
- PR without: 13.7 requests/s
```
VLLM_USE_V1=1 python benchmarks/benchmark_throughput.py --input-len 1024 --output-len 512 --num-prompts 1000 --async-engine \
--disable-log-requests [--disable-log-stat]
```

### Latency

```
============ Serving Benchmark Result ============
Metric                                  | Main       | PR with stats | PR without stats
----------------------------------------------------------------------------------------
Successful requests                     | 998        | 998           | 998
Benchmark duration (s)                  | 32.93      | 32.85         | 32.21
Total input tokens                      | 223244     | 223244        | 223244
Total generated tokens                  | 209009     | 209177        | 209029
Request throughput (req/s)              | 30.31      | 30.38         | 30.98
Output token throughput (tok/s)         | 6347.79    | 6366.81       | 6488.77
Total Token throughput (tok/s)          | 13127.92   | 13161.77      | 13418.80
---------------Time to First Token----------------
Mean TTFT (ms)                          | 3178.80    | 3134.97       | 3171.81
Median TTFT (ms)                        | 2980.97    | 3053.03       | 2960.46
P99 TTFT (ms)                           | 5466.19    | 5611.05       | 5377.22
-----Time per Output Token (excl. 1st token)------
Mean TPOT (ms)                          | 172.70     | 180.67        | 160.93
Median TPOT (ms)                        | 124.75     | 127.83        | 120.60
P99 TPOT (ms)                           | 438.38     | 599.12        | 409.44
---------------Inter-token Latency----------------
Mean ITL (ms)                           | 75.22      | 77.18         | 72.43
Median ITL (ms)                         | 36.60      | 36.04         | 36.26
P99 ITL (ms)                            | 852.84     | 908.12        | 806.74
==================================================

```

### CODE REVIEW

File: vllm/v1/engine/async_llm.py [hunk 1]
```diff
@@ -281,10 +295,13 @@ async def _run_output_handler(self):
         try:
             while True:
                 # 1) Pull EngineCoreOutput from the EngineCore.
-                outputs = await self.engine_core.get_output_async()
+                outputs: EngineCoreOutputs = (
```
Comment (robertgshaw2-neuralmagic) [inline vllm/v1/engine/async_llm.py:298]: One note: this is a blocking call. So if no work is done by the EngineCore, then this will wait forever. I wonder if we need some way to send logs if no work is happening like we have now

File: vllm/v1/engine/core.py [hunk 2]
```diff
@@ -297,21 +305,27 @@ def run_busy_loop(self):
...
+
+            self._log_stats(stats)
 
-            self._log_stats()
+    def _make_stats(self,
+                    engine_outputs: List[EngineCoreOutput]) -> EngineCoreStats:
+        return EngineCoreStats(scheduler_stats=self.scheduler.get_stats(), )
 
-    def _log_stats(self):
+    def _log_stats(self, stats: EngineCoreStats):
```
Comment (robertgshaw2-neuralmagic) [inline vllm/v1/engine/core.py:319]: Should this be removed and all logging happens in the frontend process?

File: vllm/v1/engine/async_llm.py (hunk 1 above)
Comment (rickyyx) [reply vllm/v1/engine/async_llm.py:298]: Oh - that's good to know! I wonder if that would mean we would need to have another data flow for the stats other than piggybacking on the engine outputs.

Comment (rickyyx) [reply vllm/v1/engine/async_llm.py:298]: Looks like we are already doing `profile` as another cross-process call other than adding request, I wonder if we can extend the interface by adding a `get_stats()` call to the `MPClient` in parallel with `get_output()`.

Caller could then have an async coroutine pulling stats from the core backend without being blocked by the `await get_output_async()`

**Comment by rickyyx** [comment]
So in order to support the below:
1. Stats collection should be possible when engine core is busy (or stuck)
2. Stats collection should incur minimal overheads --> happens on the frontend of the engine. 

I am proposing the below changes:
1. Introducing a `EngineStatsAgent` that lives with the `EngineCore` to collect stats from various core components and scheduler/model outputs
3. Introducing a `EngineStatsManager` class that lives on the `AsyncLLM` engine that aggregates stats from the agent.
4. The manager and agent communicates through the ZMQ sockets for async stats collection. 

```
Main Process                                  Engine Core Process
+----------------------+                      +----------------------+
|      AsyncLLM        |                     |    EngineCoreProc    |
|  +-----------------+ |                     |  +-----------------+  |
|  |EngineStatsManager| |                    |  |EngineStatsAgent |  |
|  +-----------------+ |                     |  +-----------------+  |
|                      |                     |                      |
|  [AsyncIO Event Loop]|     ZMQ Sockets     |  [Background Threads]|
|  +-----------------+ |                     |  +-----------------+  |
|  | Output Handler  |<|---------------------|->| Output Socket   |  |
|  +-----------------+ |                     |  +-----------------+  |
|                      |                     |                      |
|  +-----------------+ |                     |  +-----------------+  |
|  | Stats Handler   |<|---------------------|->| Stats Socket    |  |
|  +-----------------+ |                     |  +-----------------+  |
|                      |                     |                      |
|                      |                     |  +-----------------+  |
|                      |---------------------|->| Input Socket    |  |
|                      |                     |  +-----------------+  |
|                      |                     |                      |
|                      |                     |  [Main Thread]       |
|                      |                     |  +-----------------+  |
|                      |                     |  |   Busy Loop    |  |
|                      |                     |  +-----------------+  |
+----------------------+                      +----------------------+
```

File: vllm/v1/engine/async_llm.py [hunk 3]
```diff
@@ -295,11 +330,35 @@ async def _run_output_handler(self):
...
+        except asyncio.CancelledError:
+            logger.info("Engine shutting down.")
+            self.shutdown()
         except BaseException as e:
             logger.error(e)
             raise e
 
-    # TODO: can we eliminate these?
+    async def _run_stats_handler(self):
+        while True:
```
Comment (rickyyx) [inline vllm/v1/engine/async_llm.py:341]: This is where stats were polled from the backend.

File: vllm/v1/engine/core.py [hunk 4]
```diff
@@ -344,6 +379,18 @@ def process_input_socket(self, input_path: str):
                 request_type = type_frame.buffer
                 request_data = data_frame.buffer
 
+                if request_type == EngineCoreRequestType.STATS.value:
```
Comment (rickyyx) [inline vllm/v1/engine/core.py:379]: This is where the backend process responds. Note this is independent from the actual execution thread.

File: vllm/v1/core/scheduler.py [hunk 5]
```diff
@@ -290,6 +303,9 @@ def schedule(self) -> "SchedulerOutput":
             free_encoder_input_ids=self.encoder_cache_manager.get_freed_ids(),
         )
 
+        if self.log_stats:
```
Comment (robertgshaw2-neuralmagic) [inline vllm/v1/core/scheduler.py:306]: One thing that got us in trouble in V0 is too many loops over these types of data structures. 

I wonder if the creation of the request stats could be done in the list comprehensions used above,

File: vllm/v1/engine/async_llm.py [hunk 6]
```diff
@@ -146,6 +170,7 @@ async def add_request(
         detokenizer_req, engine_core_req = self.processor.process_inputs(
             request_id, prompt, params, arrival_time, lora_request,
             trace_headers, prompt_adapter_request, priority)
+        self.stats_manager.record_engine_input(engine_core_req)
```
Comment (robertgshaw2-neuralmagic) [inline vllm/v1/engine/async_llm.py:173]: do we need this for `LLM` as well?

Comment (robertgshaw2-neuralmagic) [reply vllm/v1/engine/async_llm.py:173]: Also, where is `arrival_time` used?

File: vllm/v1/stats/stats_manager.py [hunk 7]
```diff
@@ -0,0 +1,366 @@
...
+    def log_stats(self, stats: Stats) -> None:
+        """
+        Log the engine stats.
+        """
+
+        for stat_logger in self._stat_loggers.values():
+            stat_logger.log(stats)
+
+
+class ThreadSafeEngineStatsManager(EngineStatsManager):
```
Comment (robertgshaw2-neuralmagic) [inline vllm/v1/stats/stats_manager.py:311]: There are not any threads in the `AsyncLLM` class right now (its just using `asyncio`. So the `Threadsafe` variant is not needed.

File: vllm/v1/engine/async_llm.py [hunk 8]
```diff
@@ -74,9 +78,26 @@ def __init__(
...
+        self.stats_handler: Optional[asyncio.Task] = None
+
+        if vllm_config.observability_config.log_stats:
+            # The stats manager will be invoked from different threads:
+            # 1. stats_handler: a background busy event loop that pulls stats
+            # from the EngineCore periodically. The stats logging also happens
+            # on the background thread.
+            # 2. the main event loop when a request is added (and input
+            # processed).
+            self.stats_manager = ThreadSafeEngineStatsManager(
```
Comment (robertgshaw2-neuralmagic) [inline vllm/v1/engine/async_llm.py:97]: There are not any threads in this class (just `asyncio`) --- so we don't need a threadsafe variant.

**Review by robertgshaw2-redhat — COMMENTED** [review]
I think this design is on the right track. As discussed in slack (https://vllm-dev.slack.com/archives/C07QTE01QFQ/p1732588846571919), I think that we should remove the `EngineStatsAgent` from the `EngineCore` and instead send the stats back in `EngineCoreOutputs` each step. 

This will simplify the state management in the `EngineCore` and reduce the number of RPCs we have to worry about with the same amount of work being executed on the core busy loop as the current design.

File: vllm/v1/engine/async_llm.py (hunk 6 above)
Comment (rickyyx) [reply vllm/v1/engine/async_llm.py:173]: oh yes - the prototype was only on async llm, but it should be the same for sync llm engine.

File: vllm/v1/stats/stats_manager.py (hunk 7 above)
Comment (rickyyx) [reply vllm/v1/stats/stats_manager.py:311]: Oh right - good catch. I wonder in the future we should probably isolate the stats logger in another thread since the stats logger could be pretty expensive (e.g. the prometheus logging is quite slow when it comes to histograms).

**Comment by hmellor** [comment]
Closing as stale, feel free to re-open if you feel this PR is still relevant.
